### PR TITLE
Tinkering

### PR DESCRIPTION
**Pull request by geralddejong**
Switch bricks, then connect.  Shift-click picking.

### CODE REVIEW

File: src/application.rs [hunk 1]
```diff
@@ -168,6 +197,13 @@ impl Application {
         };
         self.user_interface.key_pressed(keycode);
     }
+
+    fn selected_face(&self) -> Option<UniqueId> {
+        let Ok([&face_id]) = self.selected_faces.iter().next_chunk() else {
```
Comment (mitchelljustin) [inline src/application.rs:202]:
```suggestion
        let Some(&face_id) = self.selected_faces.iter().next() else {
```

File: src/application.rs [hunk 2]
```diff
@@ -168,6 +197,13 @@ impl Application {
...
+            return None
+        };
+        Some(face_id)
```
Comment (mitchelljustin) [inline src/application.rs:205]:
```suggestion
        self.selected_faces.iter().next().map(|&id| id)
```

File: src/build/tenscript/build_phase.rs [hunk 3]
```diff
@@ -236,8 +231,10 @@ impl BuildPhase {
                 buds.push(Bud { face_id, forward: forward.clone(), scale_factor: *scale_factor, node })
             }
             Branch { face_nodes, rotation, alias } => {
-                let attach_to = Self::find_launch_face(&launch, &faces, fabric);
-                let brick_faces = fabric.attach_brick(alias, rotation.into(), 1.0, attach_to);
+                let launch_face = Self::find_launch_face(&launch, &faces, fabric);
+                let (base_face_id, brick_faces) = 
+                    fabric.create_brick(alias, rotation.into(), 1.0, launch_face);
+                launch_face.map(|face_id|fabric.join_faces(base_face_id, face_id));
```
Comment (mitchelljustin) [inline src/build/tenscript/build_phase.rs:237]:
```suggestion
                launch_face.map(|face_id| fabric.join_faces(base_face_id, face_id));
```

File: src/build/tenscript/shape_phase.rs [hunk 4]
```diff
@@ -180,21 +185,22 @@ impl ShapePhase {
...
+            ShapeOperation::Bouncy => Bouncy,
         }
     }
 
     fn complete_shaper(&self, fabric: &mut Fabric, Shaper { interval, alpha_face, omega_face, join, .. }: Shaper) {
+        fabric.remove_interval(interval);
         if join {
             fabric.join_faces(alpha_face, omega_face);
         }
-        fabric.remove_interval(interval);
```
Comment (mitchelljustin) [inline src/build/tenscript/shape_phase.rs:197]: Ooh, is there a story here? Was something breaking?

File: src/build/tinkerer.rs [hunk 5]
```diff
@@ -34,16 +48,38 @@ impl Tinkerer {
...
+                    self.proposed_connect = Some((base_face_id, *face_id));
+                    PendingFaceJoin
+                } else {
+                    Navigating
+                }
+            }
+            PendingFaceJoin => PendingFaceJoin,
+            Connect => {
+                if let Some(pair) = self.proposed_connect {
+                    fabric.join_faces(pair.0, pair.1);
```
Comment (mitchelljustin) [inline src/build/tinkerer.rs:69]:
```suggestion
                if let Some(pair) = self.proposed_connect {
                    fabric.join_faces(pair.0, pair.1);
```
```suggestion
                if let Some((alpha, omega)) = self.proposed_connect {
                    fabric.join_faces(alpha, omega);
```

File: src/camera.rs [hunk 6]
```diff
@@ -66,48 +70,58 @@ impl Camera {
...
     }
 
+    pub fn pick(&mut self, position: PhysicalPosition<f64>, fabric: &Fabric) {
+        let width = self.size.width / 2.0;
+        let height = self.size.height / 2.0;
+        let x = (position.x - width) / width;
+        let y = (height - position.y) / height;
+        let position = Point3::new(x as f32, y as f32, 1.0);
+        let point3d = self.mvp_matrix().invert().unwrap().transform_point(position);
+        let ray = (point3d - self.position).normalize();
```
Comment (mitchelljustin) [inline src/camera.rs:100]: Very cool

File: src/camera.rs [hunk 7]
```diff
@@ -128,48 +142,31 @@ const OPENGL_TO_WGPU_MATRIX: Matrix4<f32> = Matrix4::new(
...
-                fabric.faces.get(face_id).map(|face| {
-                    Point3::from_vec(face.midpoint(fabric))
-                })
+            Target::AroundFaces(face_set) => {
+                let midpoints = face_set
+                    .iter()
+                    .flat_map(|face_id|
+                        fabric.faces.
+                            get(face_id)
+                            .map(|face| face.midpoint(fabric)));
```
Comment (mitchelljustin) [inline src/camera.rs:164]:
```suggestion
                        fabric.faces
                            .get(face_id)
                            .map(|face| face.midpoint(fabric)));
```

File: src/camera.rs [hunk 8]
```diff
@@ -128,48 +142,31 @@ const OPENGL_TO_WGPU_MATRIX: Matrix4<f32> = Matrix4::new(
...
             Target::Origin => Some(point3(0.0, 0.0, 0.0)),
             Target::FabricMidpoint => Some(fabric.midpoint()),
-            Target::SelectedFace(face_id) => {
-                fabric.faces.get(face_id).map(|face| {
-                    Point3::from_vec(face.midpoint(fabric))
-                })
+            Target::AroundFaces(face_set) => {
+                let midpoints = face_set
+                    .iter()
+                    .flat_map(|face_id|
```
Comment (mitchelljustin) [inline src/camera.rs:161]:
```suggestion
                    .filter_map(|face_id|
```
flat_map() technically works the same, but should only be used to flatten iterables. Here you're just filtermapping Options

File: src/crucible.rs [hunk 9]
```diff
@@ -39,43 +50,46 @@ impl Default for Crucible {
...
                     if let Some(tinker_action) = tinkerer.iterate(&mut self.fabric) {
                         actions.push(tinker_action);
                     }
                 }
-                if tinkerer.is_done() {
-                    self.stage = Finished;
-                }
+            }
+            PretensingLaunch(surface_character) => {
+                actions.push(Action::Scene(SceneAction::Variant(SceneVariant::Pretensing)));
```
Comment (mitchelljustin) [inline src/crucible.rs:91]: That's some enum nesting. Not that that's a problem

File: src/crucible.rs [hunk 10]
```diff
@@ -98,27 +112,53 @@ impl Crucible {
...
                 };
-                let spin = self.fabric.face(face_id).spin.opposite();
-                let face_alias = FaceAlias::single("Single") + &spin.into_alias();
-                tinkerer.add_brick(face_alias, face_id);
+                tinkerer.propose_brick(brick_on_face);
             }
-            CrucibleAction::SetSpeed(iterations_per_frame) => {
+            ConnectBrick => {
+                let Tinkering(tinkerer) = &mut self.stage else {
+                    panic!("cannot add brick unless tinkering");
```
Comment (mitchelljustin) [inline src/crucible.rs:132]: Can this be an unreachable!()?

File: src/crucible.rs [hunk 11]
```diff
@@ -98,27 +112,53 @@ impl Crucible {
...
+                    tinkerer.join_faces(a, b);
+                }
+            }
+            SetSpeed(iterations_per_frame) => {
                 self.iterations_per_frame = iterations_per_frame;
             }
+            InitiateRevert => {
+                let Tinkering(tinkerer) = &mut self.stage else {
+                    panic!("cannot revert unless tinkering");
+                };
```
Comment (mitchelljustin) [inline src/crucible.rs:150]: Should extract this to a method

File: src/fabric/pretenser.rs [hunk 12]
```diff
@@ -21,13 +21,15 @@ pub struct Pretenser {
 }
 
 impl Pretenser {
-    pub fn new(pretenst_factor: f32) -> Self {
+    pub fn new(pretenst_factor: f32, surface_character: SurfaceCharacter) -> Self {
+        let mut physics = AIR_GRAVITY;
+        physics.surface_character = surface_character;
```
Comment (mitchelljustin) [inline src/fabric/pretenser.rs:26]:
```suggestion
        let physics = Physics {
            surface_character,
            ..AIR_GRAVITY
        };
```

Comment (mitchelljustin) [reply src/fabric/pretenser.rs:26]: Could even then inline `physics` into the returned Pretenser

**Review by mitchelljustin — APPROVED** [review]

File: src/application.rs (hunk 2 above)
Comment (geralddejong) [reply src/application.rs:205]: What if there are two?  You just get the first, right?  I don't want any if there are two.

File: src/build/tenscript/shape_phase.rs (hunk 4 above)
Comment (geralddejong) [reply src/build/tenscript/shape_phase.rs:197]: No.  It actually doesn't matter.

File: src/crucible.rs (hunk 10 above)
Comment (geralddejong) [reply src/crucible.rs]: I've now delegated the tinkerer actions all in one.

File: src/application.rs (hunk 2 above)
Comment (mitchelljustin) [reply src/application.rs:205]: Then neither solution works. next_chunk() will still give you the next one element, whether or not there are more.

**Pull request merged**